ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/PullQueryStreamWriter.java

```diff
@@ -0,0 +1,215 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server.resources.streaming;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.Lists;
+import io.confluent.ksql.api.server.StreamingOutput;
+import io.confluent.ksql.execution.streams.materialization.Locator.KsqlNode;
+import io.confluent.ksql.physical.pull.PullQueryResult;
+import io.confluent.ksql.physical.pull.PullQueryRow;
+import io.confluent.ksql.query.PullQueryQueue;
+import io.confluent.ksql.rest.Errors;
+import io.confluent.ksql.rest.entity.KsqlHostInfoEntity;
+import io.confluent.ksql.rest.entity.StreamedRow;
+import io.confluent.ksql.util.KsqlException;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.time.Clock;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("checkstyle:CyclomaticComplexity")
+public class PullQueryStreamWriter implements StreamingOutput {
+  private static final Logger LOG = LoggerFactory.getLogger(PullQueryStreamWriter.class);
+
+  private static final int FLUSH_SIZE_BYTES = 50 * 1024;
+  private static final long MAX_FLUSH_MS = 30;
+
+  private final long disconnectCheckInterval;
+  private final PullQueryQueue pullQueryQueue;
+  private final Clock clock;
+  private final PullQueryResult result;
+  private final ObjectMapper objectMapper;
+  private volatile boolean limitReached = false;
+  private volatile boolean connectionClosed;
+  private volatile Throwable pullQueryException;
+  private boolean closed;
+  private boolean sentAtLeastOneRow = false;
+
+  PullQueryStreamWriter(
+      final PullQueryResult result,
+      final long disconnectCheckInterval,
+      final ObjectMapper objectMapper,
+      final PullQueryQueue pullQueryQueue,
+      final Clock clock,
+      final CompletableFuture<Void> connectionClosedFuture
+  ) {
+    this.result = Objects.requireNonNull(result, "result");
+    this.objectMapper = Objects.requireNonNull(objectMapper, "objectMapper");
+    this.disconnectCheckInterval = disconnectCheckInterval;
+    this.pullQueryQueue = Objects.requireNonNull(pullQueryQueue, "pullQueryQueue");
+    this.clock = Objects.requireNonNull(clock, "clock");
+    connectionClosedFuture.thenAccept(v -> connectionClosed = true);
+    result.onException(t -> pullQueryException = t);
+    result.onCompletion(v -> limitReached = true);
+  }
+
+  @Override
+  public void write(final OutputStream output) {
```

Comment:
there's a lot of different levels going on and wrapping my head around passing around different heads/string buffers and flushing in the middle (etc...) is a struggle. At a minimum, I think we can do with inline comments and javadocing (including the private methods), though I challenge you to take another stab and see if you can refactor this to make it cleaner and reduce the cyclomatic complexity (this is one of those times where I think checkstyle got it right, it needs improvement).

There's a lot of complexity in creating proper JSON (commas in the right places, the header row being first, etc...). I wonder if there's a library that we can use to make that simpler.

Reply:
💯 this is _way_ cleaner and easier for me to reason about. Thanks for the heavy refactoring!